design-documents/service-isolation/authentication-authorization.md

```diff
@@ -0,0 +1,146 @@
+to `Auth Service`, which can automatically assign API to existing roles based on default rules. The token validation includes
+not only signature verification but authorization.
+
+### Services Deployment
+
+`Auth Service` stores allowed operations for roles and, during deployment, each service should send a list of available
+operations to `Auth Service`.
+
+![Services Deployment](services-deployment.png)
+
+### Service-to-Service Communication
+
+As each service should be agnostic to a source of a request (in general, for a service it does not matter who sends
+a request) service-to-service communication also requires a token. The next diagram describes a possible flow:
+
+![Service to Service](service-to-service.png)
+
+Before making a call to the needed service, the first service makes a call to `Auth Service` to retrieve token for the future
+communication. The second service makes a call for token validation in the same way as for other types of requests, so
+for the second service, it does not matter who sends original request.
+
+### Guest-to-Service Communication
+
+As users do not allowed making requests to services directly, a communication for guest customers also should use tokens.
+The main difference that `Auth Token` issues a token based on guest role:
+
+![Guest to Service](guest-to-service.png)
+
+### OAuth 2.0
+
+As `Auth Service` is the single entry for authentication & authorization it should store roles and privileges for all
+services and generate an access token, this schema follows to OAuth 2.0 framework principles - `Auth Service` has a role
+as `Authorization Server` and it a centralized storage of privileges for `Resource Servers`.
+
+The usage of OAuth 2.0 framework has the next benefits:
+* Defines authorization grant flows for different kinds of a communication
+* Defines the rules for a token obtaining, issuing, refreshing
+* Service-to-service authentication can be based on Client Credentials Grant ([CCG](https://tools.ietf.org/html/rfc6749#section-4.4))
+* JWT could be used for a client authentication and authorization ([RFC7523](https://tools.ietf.org/html/rfc7523))
+* Defines the standard for the response format, error codes
+
+## Open Questions
+
+ - Anyway, each service should check if a user is allowed to perform an action (for example, if the current user is allowed
+ editing profile details because only service knows the profile owner)
+ - Search WEB API operations like `getList` should be filtered by a request initiator (a customer can retrieve only his
+ orders, a guest does not have customer ID, admin can get all orders, a company user - according to the allowed hierarchy)
+ - Roles with hierarchy probably might use [Open Policy Agent](https://www.openpolicyagent.org/)
+
+## Implementation Approach
+
+On the first iteration, BFF/Monolith could be used as `Auth Service` and as a single point for roles & privileges storage.
```

Comment:
How authorization will be enabled on each service and monolith? Do we enable authorization for all requests or only for post operations and ones that contain sensitive data (I think this would be better)?

Reply:
I think for GET requests authorization also should be enabled because service should know if request initiator is allowed to request such data.

Reply:
I'm not sure we need to perform authorization/authentication for all get requests. Number of calls and ability to efficiently cache requests is major concern right now. Caching in the application will have much lover performance and higher overhead than caching by Varnish. I think we don't need to perform authorization for GET requests that don't expose sensitive data.

Reply:
I partially agree but, for example, order items details are not sensitive data but probably we won't provide an ability to expose ordered items from one customer to another.

Reply:
I would consider order items sensitive data as it's part of order data. Not sure we should have API to get only order items data. We can cache for instance product, reviews and tax data.

Reply:
Actually, now we have WEB API endpoint to get order items. In general, order item just contains product details (which are not sensitive). Mentioned example just to show that not always we can strictly divide some data to sensitive and not.